data_analysis/correction_framework.py

```diff
@@ -0,0 +1,247 @@
+                data[4].float,
+                data[5],
+            )
+            if (
+                m_rec.local_time > self.sunrise
+            ):  # unsure if this will work or if a special time comp function is needed
+                m_file.readline()
+                break
+            self.min_records.append(m_rec)
+            transtr = m_file.readline()
+        m_file.close()
+
+        self.valid_for_use = True
+        self.exclusion_reason = "None"
+
+    # mark data as invalid based on various night conditions
+    def grade_data(self):
+        # taking the average night to be ten hours (600 minutes), if more than three hours (180 minutes) are missing,
+        # invalidate the data based on insufficient records
+        if len(self.min_records) < 420:
+            self.valid_for_use = False
+            self.exclusion_reason = "Insufficient Records for Night"
+
+        # if the moon would have been obscured by bad weather for more then three hours, invalidate based on weather conditions
+        # use something other than python-weather?
+
+    # given a function on a float, updates all temperature, frequency, or voltage values
+    def apply_correction(self, field_to_update: str, funct):
+        if field_to_update == "TEMP":
+            for rec in self.min_records:
+                rec.temperature = funct(rec.temperature)
+        if field_to_update == "FREQ":
+            for rec in self.min_records:
+                rec.frequency = funct(rec.frequency)
+        if field_to_update == "VOLT":
+            for rec in self.min_records:
+                rec.voltage = funct(rec.voltage)
+        else:
+            print("error: field not recognized/field should not be updated")
+
+    # export night record to a new file
+    def rec_export(self, filename: str):
+        new_file = open(filename, "x")
+        if self.valid_for_use == False:
+            new_file.write("INVALID\n")
+            new_file.write(self.exclusion_reason)
+        else:
+            # night date -- 2023-01-23 06:00:00.992
+            new_file.write(self.night_date.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # morning date
+            new_file.write(self.morning_date.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # sunrise
+            new_file.write(self.sunrise.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # sunset
+            new_file.write(self.sunset.strptime("%Y-%m-%d %H:%M:%S.%f"))
+            new_file.write("\n")
+            # twilight
+            new_file.write(
+                self.astronomical_twilight.strptime("%Y-%m-%d %H:%M:%S.%f")
+            )
+            new_file.write("\n")
+            # illumination
+            new_file.write(str(self.moon_illumination))
+            new_file.write("\n")
+            # weather
+            # unimplemented
+
+            for rec in self.min_records:
+                new_file.write(
```

Comment:
Instead of using `write()` several times, you can just submit all of your lines at once using the `writelines()` function. An example of that is here: https://www.geeksforgeeks.org/difference-between-write-and-writelines-function-in-python/#tablist2-tab1

Reply:
For the minute_records, I chose to read and write them one at a time as opposed to all at once because there are likely to be a lot of them for any given night, ideally around 600 per night (~1 per minute of ~10 hours), and it seemed more efficient to just write them as they're read instead of creating an array of a few hundred elements or a super long string of them to write all at once. Though if writing them all at once would be more efficient, let me know and I'll swap it to that no problem.

Reply:
I don't think there's much to worry about with the scope being 600 elements. Provided you aren't doing any complicated sorts, etc. either way is likely more than fine.